docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+  providerConfig:
+    apiVersion: aws.cloud.gardener.cloud/v1alpha1
+    kind: WorkerPoolConfig
+    ami: ami-d0dcef3b
+    machineType: m4.large
+    volumeType: gp2
+    volumeSize: 20Gi
+    zones:
+    - eu-west-1a
+  region: eu-west-1
+  secretRef:
+    name: my-aws-credentials
+  minimum: 2
+  maximum: 2
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+```
+
+#### Shoot state
+
+In order to be able to move the control plane of a Shoot between Seed clusters (e.g., if a Seed cluster is not available anymore or entirely broken) Gardener must store some non-reconstructable state, potentially also the state written by the controllers.
+Gardener watches over the extension CRDs and copies the `.status.state` in a `ShootState` resource in the Garden cluster.
+Any change to the status written by the respective CRD-controllers must be immediately reflected in the `ShootState` resource.
+The contract between Gardener and those controllers is: **Every controller must be capable of reconstructing its own environment based on both the state it has written before and on the real world's conditions/state.**
+
+```yaml
+---
+apiVersion: gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: shoot--core--aws-01
+shootRef:
+  name: aws-01
+  project: core
+state:
+  secrets:
+  - name: ca
+    data: ...
+  - name: kube-apiserver-cert
+    data: ...
+  resources:
+  - kind: DNS
+    name: record-1
+    state: <copied-state-of-dns-crd>
+  - kind: Infrastructure
+    name: networks
+    state: <copied-state-of-infrastructure-crd>
+  ...
+  <other fields required to keep track of>
+```
+
+We cannot assume that Gardener is always online to observe the most recent states the controllers have written to their resources.
+Consequently, the information stored here must not be used as "single point of truth", but the controllers must potentially check the real world's status to reconstruct themselves.
+However, this must anyway be part of their normal reconciliation logic and is a general best practice of writing Kubernetes controllers.
+
+#### Shoot health checks/conditions
+
+Some of the existing conditions already contain specific code which shall be simplified as well.
+All of the CRDs described above have a `.status.conditions` field to which the controllers may write relevant health information of their function area.
+Gardener will pick them up and copy them over to the Shoots `.status.conditions`.
+
+#### Reconciliation flow
+
+We are now examining the current Shoot creation/reconciliation flow and describe roughly how it would change when applying this proposal:
+
+| Operation | Description |
+|-----------|-------------|
+| botanist.DeployNamespace | Gardener creates the namespace for the Shoot in the Seed cluster. |
+| botanist.DeployKubeAPIServerService | Gardener creates a Service of type `LoadBalancer` in the Seed.<br>AWS Botanist registers a Mutating Webhook and adds its AWS-specific annotation. |
```

Comment:
What if the `MutatingWebHook` is down when this CRD is created? Isn't a *proper* controller a more reliable approach?

Reply:
> why is this generic key associated to an AWS dependent annotation ?

Because right now (current implementation) there is an AWS-specific annotation on these `Service`s. Botanists might put their cloud specific information into the annotation (or might even need to change the spec), this was just an "real-world" example.

Reply:
> What if the MutatingWebHook is down when this CRD is created? Isn't a proper controller a more reliable approach?

Then the request will fail. Do you suggest to even let that `Service`  to be created by someone else? It would be harder to create DNS records, etc., Gardener needs to know the API server ingress at some places. I assume it makes our lifes more complicated, at least in the beginning.

Reply:
My comments were more in general rather than just this `Service`. It is one thing to use `MutatingWebhooks` for customisation. But we are proposing them for the basic implementation of the flow (for this `Service`, `ETCD`, many other control-plane components etc.) This would couple the error handling for these things with the Gardener reconciliation cycle.

Shouldn't decoupling such basic error handling from Gardener reconciliation be one of the goals of this extensibility proposal?

If CRDs are used instead then the controller can perform the required error handling by itself independent of the Gardener reconciliation.

If the concern is that this would increase the number of CRDs too much then perhaps we can have a shared ShootControlPlaneCRD which multiple controllers can watch and react to.

Reply:
OK, understood. Thing is that we still want to keep control over the flow on Gardener side, and basic things that are the same in all environments (e.g., the need to have an etcd) would stay in Gardener's responsibility. So far, we have highly homogeneous clusters and as soon as we move too many basics out (load balancer service, etcd, control plane, ...) then we might not be able to preserve that.
We decided that those parts that are/look the same in all environments (aside from some small configuration changes) shall be handled by webhooks, whereas those parts which are more specific must be handled by dedicated controllers.
Hope that makes sense. However, that doesn't mean that we can't move into the direction of replacing more webhooks by dedicated controllers in the future. It's just that the whole extraction is a tough challenge on its own and we should be cautious not to make Gardener loosing its power to maintain/repair/manage the clusters reliably.

Reply:
Fair enough. But I think we should at least add a couple of sentences about the availability of the `MutatingWebhooks` and the implications if they are unavailable temporarily.

Reply:
What is the argument against using the current approach of Gardener deploying the whole thing for the non-cloud-provider-specific components? For example, why not let Gardener continue to deploy the whole ETCD `StatefulSet` (sidecar and all)?

To me, either a CRD approach or Gardener deploying the whole thing makes sense. Relying on `MutatingWebhooks` for this seems to be a middle-path that neither seems to gain in abstraction nor extensibility.

Reply:
The argument is that the backup-restore sidecar needs cloud-specific configuration that Gardener doesn't know.

Reply:
Understood. Thanks.